slides/C-concurrency-parallelism.md

```diff
@@ -0,0 +1,613 @@
+- our operation has a neutral value `HashMap::default()`: `0 • x = x • 0 = x`
+- therefore we can split the computation `a • b • c • d = (0 • a • b) • (0 • c • d)`
+- an associative operation with a neutral value is called a "monoid"
+
+```rust
+// for each word, how often it occurs across all documents
+documents
+    .par_iter()
+    .map(|document| count_words(document))
+    .reduce(HashMap::default, combine_documents);
+```
+
+- this idea means each thread can start accumulating values
+
+
+---
+layout: default
+---
+
+# So far
+
+- Rayon makes data-parallel programming in rust extremely convenient
+
+
+---
+theme: default
+class: text-center
+highlighter: shiki
+lineNumbers: true
+info: "Rust - X: Y"
+drawings:
+  persist: false
+fonts:
+  mono: Fira Mono
+layout: cover
+title: 'Rust - X: Y'
+---
+# Fearless concurrency
+thread-based concurrency in rust
+
+
+---
+layout: default
+---
+
+# Fearless concurrency
+
+```rust
+use std::thread;
+
+fn main() {
+    thread::spawn(f);
+    thread::spawn(f);
+
+    println!("Hello from the main thread.");
+}
+
+fn f() {
+    println!("Hello from another thread!");
+
+    let id = thread::current().id();
+    println!("This is my thread id: {id:?}");
+}
+```
+
+- A process can spawn multiple threads of execution. These run concurrently (and may run in parallel)
+- Question: what is the output of this program?
+
+---
+layout: default
+---
+
+# Expected output
+
+maybe
+
+```
+Hello from another thread!
+This is my thread id: ThreadId(411)
+Hello from another thread!
+This is my thread id: ThreadId(412)
+Hello from the main thread.
+```
+
+or
+
+```
+Hello from another thread!
+Hello from another thread!
+This is my thread id: ThreadId(411)
+This is my thread id: ThreadId(412)
+Hello from the main thread.
+```
+
+---
+layout: default
+---
+
+# Expected output
+
+but most likely
+
+```
+Hello from the main thread.
+```
+
+The process exits when the main thread is done!
+
+
+- `.join()` can be used to block the main thread until the child is done
+
+```rust
+fn main() {
+    let t1 = thread::spawn(f);
+    let t2 = thread::spawn(f);
+
+    println!("Hello from the main thread.");
+
+    t1.join().unwrap();
+    t2.join().unwrap();
+}
+```
+
+- `.join()` turns a panic in the thread into an `Err`
+
+---
+layout: default
+---
+
+# Thread lifetime
+
+- a more typical example
+
+```rust
+let numbers = Vec::from_iter(0..=1000);
+
+let t = thread::spawn(move || {
+    let len = numbers.len();
+    let sum = numbers.iter().sum::<usize>();
+    sum / len
+});
+
+let average = t.join().unwrap();
+
+println!("average: {average}");
+```
+
+- `numbers` must be `move`d into the closure!
+
+---
+layout: default
+---
+
+# Thread lifetime
+
+- otherwise `numbers` might be dropped while the thread is still using it!
+
+```rust
+let numbers = Vec::from_iter(0..=1000);
+
+let t = thread::spawn(|| {
+    let len = numbers.len();
+    let sum = numbers.iter().sum::<usize>();
+    sum / len
+});
+
+drop(numbers); // oh no
```

Comment:
Doesn't the Rust type system slap you on the wrist here? The way the slide is made with the "oh no") makes it look like you are on your own, as you would in C.